collection+xattrs.go

```diff
@@ -304,113 +386,197 @@ func (c *Collection) getRawWithXattr(key string, xattrKey string, userXattrKey s
 	return
 }
 
-// Replaces a doc's body and/or an xattr.
+// Delete a document's body and an xattr simultaneously.
+func (c *Collection) _deleteBodyAndXattr(
+	key string,
+	xattrKey string,
+	bodyMustExist bool,
+) error {
+	err := c.withNewCas(func(txn *sql.Tx, newCas CAS) (*event, error) {
+		e := &event{
+			key:        key,
+			cas:        newCas,
+			isDeletion: true,
+		}
+		var bodyExists bool
+		row := txn.QueryRow(`SELECT xattrs, value NOT NULL FROM documents WHERE collection=?1 AND key=?2`, c.id, key)
+		err := scan(row, &e.xattrs, &bodyExists)
+		if err != nil {
+			return nil, remapKeyError(err, key)
+		} else if bodyMustExist && !bodyExists {
+			return nil, sgbucket.MissingError{Key: key}
+		} else if e.xattrs, err = removeXattrs(e.xattrs, xattrKey); err != nil {
+			return nil, err
+		}
+		_, err = txn.Exec(`UPDATE documents SET value=null, xattrs=?1, cas=?2 WHERE collection=?3 AND key=?4`, e.xattrs, newCas, c.id, key)
+		return e, err
+	})
+	return err
+}
+
+// Option flags for writeWithXattr
+type writeXattrOpts uint8
+
+const (
+	xInsertDoc     = writeXattrOpts(1 << iota) // Create new doc; fail if it already exists
+	xInsertXattr                               // Fail if `xattr` already exists
+	xPreserveXattr                             // Leave xattr value alone, just expand macros
+	xNoOpts        = writeXattrOpts(0)         // Default behavior
+)
+
+// Swiss Army knife method for modifying a document xattr, with or without changing the body.
 func (c *Collection) writeWithXattr(
 	key string, // doc key
-	val []byte, // document body; nil means no change
-	xattr sgbucket.Xattr, // xattr; nil value means delete
+	val *payload, // if non-nil, updates doc body; a nil payload means delete
+	xattrKey string, // xattr key
+	xattrVal payload, // xattr value; a nil payload deletes the xattr
 	ifCas *CAS, // if non-nil, must match current CAS; 0 for insert
-	exp Exp, // expiration
-	isDelete bool, // ignored: I have no idea what exactly it does...
-	deleteBody bool, // if true, delete the doc body
-	xattrIsUser bool, // true if this is a user xattr
+	exp *Exp, // if non-nil, sets expiration to this value
+	opts writeXattrOpts, // option flags
```

Comment:
is it worth adding validation so that `preserveXattr` and `insertXattr` aren't set simultaneously?

Reply:
It was too noisy in practice (happens a lot in normal use with a `MissingError`) so I commented it out for the time being intending to downgrade it or maybe just ignore certain errors.

I'll re-enable it but at `debug` level.

Reply:
It's unlikely to happen since only this source file uses those flags, but it's cheap to put in a check.